network/local.go

```diff
@@ -0,0 +1,459 @@
+package network
+
+import (
+	"errors"
+	"fmt"
+	"golang.org/x/net/context"
+	"reflect"
+	"sync"
+	"time"
+)
+
+// NewLocalRouter returns a fresh router which uses local connections. It takes
+// the default context.
+func NewLocalRouter(sid *ServerIdentity) (*Router, error) {
+	return NewLocalRouterWithContext(defaultLocalContext, sid)
+}
+
+// NewLocalRouterWithContext is the same as NewLocalRouter but takes a specific
+// LocalContext. This is useful to run parallel different local overlays.
+func NewLocalRouterWithContext(ctx *LocalContext, sid *ServerIdentity) (*Router, error) {
+	h, err := NewLocalHostWithContext(ctx, sid.Address)
+	if err != nil {
+		return nil, err
+	}
+	return NewRouter(sid, h), nil
+}
+
+// LocalContext keeps reference to all opened local connections
+// It also keeps tracks of who is "listening", so it's possible to mimics
+// Conn & Listener.
+type LocalContext struct {
+	// queues maps a remote endpoint to its packet queue. It's the main
+	//stucture used to communicate.
+	queues map[endpoint]*connQueue
+	sync.Mutex
+	listening map[Address]func(Conn)
+
+	baseUID uint64
+}
+
+// NewLocalContext returns a fresh new context that can be used by LocalConn,
+// LocalListener & LocalHost.
+func NewLocalContext() *LocalContext {
+	return &LocalContext{
+		queues:    make(map[endpoint]*connQueue),
+		listening: make(map[Address]func(Conn)),
+	}
+}
+
+var defaultLocalContext = NewLocalContext()
+
+// endpoint represents one endpoint of a connection.
+type endpoint struct {
+	addr Address
+	// uid is a unique identifier of the remote endpoint
+	// it's unique  for each direction:
+	// 127.0.0.1:2000 -> 127.0.0.1:2000 => 14
+	// 127.0.0.1:2000 <- 127.0.0.1:2000 => 15
+	uid uint64
+}
+
+// LocalReset reset the whole map of connections + listener so it is like
+// a fresh defaultLocalContext.
+func LocalReset() {
+	defaultLocalContext = NewLocalContext()
+
+}
+
+// IsListening returns true if the remote address is listening "virtually"
+func (ccc *LocalContext) IsListening(remote Address) bool {
+	ccc.Lock()
+	defer ccc.Unlock()
+	_, ok := ccc.listening[remote]
+	return ok
+}
+
+// Listening put the address as "listening" mode. If a user connects to this
+// addr, this function will be called.
+func (ccc *LocalContext) Listening(addr Address, fn func(Conn)) {
```

Comment:
> , it's also setting the fact that this address is now currently listening

So, by putting some callback into the `listening` map the `LocalContext` is set into a "I'm OK to accept connections and listen" mode? So it is merely an initialization that needs to happen before it can be used to actually listen? If yes, I propose: `InitListening`, `IsListening` and `ResetListening`?